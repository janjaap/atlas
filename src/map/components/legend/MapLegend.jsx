import React from 'react';
import PropTypes from 'prop-types';

import { Checkbox } from '../../../shared/components/checkbox';
import RemoveIcon from '../../../../public/images/icon-cross.svg';

import './_map-legend.scss';

const isAuthorised = (layer, user) => (
  !layer.authScope || (user.authenticated && user.scopes.includes(layer.authScope))
);

const isInsideZoomLevel = (layer, zoomLevel) => (
  zoomLevel >= layer.minZoom && zoomLevel <= layer.maxZoom
);

class MapLegend extends React.Component {
  static constructLegendIconUrl(mapLayer, legendItem) {
    if (legendItem.iconUrl) {
      return legendItem.iconUrl;
    }
    return [
      '//',
      process.env.NODE_ENV !== 'production' ? 'acc.map.data.amsterdam.nl' : 'map.data.amsterdam.nl',
      `${mapLayer.url}&`,
      'request=GetLegendGraphic&',
      'sld_version=1.1.0&',
      `layer=${legendItem.layer || mapLayer.layers[0]}&`,
      'format=image/svg%2Bxml&',
      `rule=${encodeURIComponent(legendItem.title)}`
    ].join('');
  }

  static mapLayersLegendItemsToIds(mapLayer) {
    return [
      mapLayer.id,
      ...mapLayer.legendItems.map((legendItem) => legendItem.id)
    ].filter((mapLayerId) => !!mapLayerId);
  }

  determineLayerVisibility(mapLayer) {
    return this.props.overlays.some((overlay) => [
      { id: mapLayer.id },
      ...mapLayer.legendItems
    ].some((legendItem) => overlay.id === legendItem.id && overlay.isVisible));
  }

  determineLegendItemVisibility(legendItem) {
    return this.props.overlays.some((overlay) => overlay.id === legendItem.id && overlay.isVisible);
  }

  toggleLayer(mapLayer) {
    MapLegend.mapLayersLegendItemsToIds(mapLayer).forEach((mapLayerId) =>
      this.props.onLayerToggle(mapLayerId));
  }

  toggleLayerVisibility(mapLayer) {
    const isVisible = this.determineLayerVisibility(mapLayer);
    MapLegend.mapLayersLegendItemsToIds(mapLayer).forEach((mapLayerId) =>
      this.props.onLayerVisibilityToggle(mapLayerId, !isVisible));
  }

  render() {
    const { activeMapLayers, onLayerVisibilityToggle, user, zoomLevel } = this.props;

    return (
      <div>
        <h3 className="u-sr-only">Actieve kaartlagen</h3>
        <ul className="map-legend">
          {activeMapLayers.map((mapLayer) => (
            <li
              className={`
                map-legend__map-layer
                map-legend__map-layer--${this.determineLayerVisibility(mapLayer) ? 'visible' : 'hidden'}
              `}
              key={mapLayer.title}
            >
              <div
                className={`
                  map-legend__category
                  map-legend__category--${mapLayer.legendItems.some((legendItem) => legendItem.selectable) ? '' : 'un'}selectable-legend
                `}
              >
                <Checkbox
                  checked={() => this.determineLayerVisibility(mapLayer)}
                  name={mapLayer.title}
                  onChange={() => this.toggleLayerVisibility(mapLayer)}
                />
                <h4 className="map-legend__category-title">{mapLayer.title}</h4>
                <button
                  className="map-legend__toggle map-legend__toggle--remove"
                  onClick={() => this.toggleLayer(mapLayer)}
                >
                  <RemoveIcon />
                </button>
              </div>
              {!isAuthorised(mapLayer, user) && (
                <div className="map-legend__notification">
                  <span>Zichtbaar na inloggen</span>
                </div>
              )}
              {isAuthorised(mapLayer, user) && !isInsideZoomLevel(mapLayer, zoomLevel) && (
                <div className="map-legend__notification">
                  <span>Zichtbaar bij verder {zoomLevel < mapLayer.minZoom ? 'inzoomen' : 'uitzoomen'}</span>
                </div>
              )}
              {isAuthorised(mapLayer, user) && isInsideZoomLevel(mapLayer, zoomLevel) && (
                <ul className="map-legend__items">
                  {mapLayer.legendItems.map((legendItem) => (
                    <li
                      className="map-legend__item"
                      key={legendItem.title}
                    >
                      {legendItem.selectable && (
                        <Checkbox
                          checked={() => this.determineLegendItemVisibility(legendItem)}
                          name={legendItem.title}
                          onChange={() => onLayerVisibilityToggle(legendItem.id)}
                        />
                      )}
                      <div className={`
                        map-legend__image
                        map-legend__image--${legendItem.selectable ? 'selectable' : 'not-selectable'}
                      `}
                      >
                        <img
                          alt=""
                          src={MapLegend.constructLegendIconUrl(mapLayer, legendItem)}
                        />
                      </div>
                      <span className="map-legend__title">{legendItem.title}</span>
                    </li>
                  ))}
                </ul>
              )}
            </li>
          ))}
        </ul>
      </div>
    );
  }
}

MapLegend.propTypes = {
  activeMapLayers: PropTypes.array, // eslint-disable-line
  onLayerToggle: PropTypes.func, // eslint-disable-line
  onLayerVisibilityToggle: PropTypes.func, // eslint-disable-line
  overlays: PropTypes.array, // eslint-disable-line
  user: PropTypes.object, // eslint-disable-line
  zoomLevel: PropTypes.number.isRequired
};

export default MapLegend;
