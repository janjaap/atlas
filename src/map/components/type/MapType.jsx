import React from 'react';
import PropTypes from 'prop-types';

import { SelectButton } from '../../../shared/components/select-button';

import './_map-type.scss';

function hasLayer(activeBaseLayer, baseLayers) {
  return baseLayers && baseLayers.length > 0 &&
    baseLayers.some((layer) => layer.value === activeBaseLayer);
}

const MapType = ({ activeBaseLayer, baseLayers, onBaseLayerToggle }) => (
  <section className="map-type">
    <h3 className="map-type__title">Achtergrond</h3>

    <SelectButton
      className="map-type__select map-type__select--topography"
      handleChange={onBaseLayerToggle}
      icon={<span className="icon icon--topography" />}
      isDisabled={!hasLayer(activeBaseLayer, baseLayers.topography)}
      name="topography"
      options={baseLayers.topography}
      value={hasLayer(activeBaseLayer, baseLayers.topography) ? activeBaseLayer : null}
    />

    <SelectButton
      className="map-type__select map-type__select--aerial"
      handleChange={onBaseLayerToggle}
      icon={<span className="icon icon--aerial" />}
      isDisabled={!hasLayer(activeBaseLayer, baseLayers.aerial)}
      name="aerial"
      options={baseLayers.aerial}
      value={hasLayer(activeBaseLayer, baseLayers.aerial) ? activeBaseLayer : null}
    />
  </section>
);

MapType.propTypes = {
  activeBaseLayer: PropTypes.string.isRequired,
  baseLayers: PropTypes.object, // eslint-disable-line
  onBaseLayerToggle: PropTypes.func.isRequired
};

export default MapType;
