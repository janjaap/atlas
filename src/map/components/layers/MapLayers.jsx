import React from 'react';
import PropTypes from 'prop-types';

import MapLegend from '../legend/MapLegend';
import AddIcon from '../../../../public/images/icon-plus.svg';
import RemoveIcon from '../../../../public/images/icon-cross.svg';

import './_map-layers.scss';

const getCategories = (layers) => (
  layers.filter((layer) => layer.category).map((layer) => layer.category)
);

const MapLayers = (props) => (
  <div className="map-layers">
    <h3 className="u-sr-only">Beschikbare kaartlagen</h3>
    <ul>
      {[...new Set(getCategories(props.layers))]
        .map((category) => (
          <li className="map-layers__category" key={category}>
            <h4 className="map-layers__category-text">{category}</h4>
            <ul>
              {props.layers
                .filter((layer) => layer.category === category)
                .map((layer) => (
                  <li
                    className={`
                      map-layers__title
                      map-layers__title--${props.activeMapLayers.some((mapLayer) => layer.title === mapLayer.title) ? 'active' : 'inactive'}
                    `}
                    key={layer.title}
                  >
                    <button onClick={() => {
                      MapLegend.mapLayersLegendItemsToIds(layer).forEach((mapLayerId) =>
                        props.onLayerToggle(mapLayerId));
                    }}
                    >
                      <span className="map-layers__toggle-title">
                        {layer.title}
                      </span>
                      <span className="map-layers__toggle map-layers__toggle--remove">
                        <RemoveIcon />
                      </span>
                      <span className="map-layers__toggle map-layers__toggle--add">
                        <AddIcon />
                      </span>
                    </button>
                  </li>
              ))}
            </ul>
          </li>
      ))}
    </ul>
  </div>
);

MapLayers.propTypes = {
  activeMapLayers: PropTypes.array, // eslint-disable-line
  layers: PropTypes.array, // eslint-disable-line
  onLayerToggle: PropTypes.func.isRequired // eslint-disable-line react/no-unused-prop-types
};

export default MapLayers;
