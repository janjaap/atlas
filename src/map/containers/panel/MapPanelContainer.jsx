import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { getMapBaseLayers, setMapBaseLayer } from '../../ducks/base-layers/map-base-layers';
import { toggleMapOverlay, toggleMapOverlayVisibility } from '../../ducks/overlays/map-overlays';
import { getMapLayers, selectActiveMapLayers } from '../../ducks/layers/map-layers';
import { toggleMapPanel, toggleMapPanelHandle } from '../../../shared/ducks/ui/ui';
import MapLayers from '../../components/layers/MapLayers';
import MapLegend from '../../components/legend/MapLegend';
import MapPanelHandle from '../../components/panel-handle/MapPanelHandle';
import MapType from '../../components/type/MapType';
import CollapseIcon from '../../../../public/images/icon-arrow-down.svg';
import ExpandIcon from '../../../../public/images/icon-arrow-up.svg';
import MapLayersIcon from '../../../../public/images/icon-map-layers.svg';

const mapStateToProps = (state) => ({
  activeBaseLayer: state.map.baseLayer,
  activeMapLayers: selectActiveMapLayers(state),
  atlas: state.atlas,
  isMapPanelVisible: state.ui.isMapPanelVisible,
  isMapLayersVisible: state.ui.isMapLayersVisible,
  isEachOverlayInvisible: state.map.overlays.every((overlay) => overlay.isVisible),
  isMapPanelHandleVisible: !state.map.overlays.length || state.ui.isMapPanelHandleVisible,
  mapBaseLayers: state.mapBaseLayers,
  mapLayers: state.mapLayers,
  overlays: state.map.overlays,
  zoomLevel: state.map.zoom,
  user: state.user
});

// TODO: Add method that checks whether layer is active and toggles accordingly
const mapDispatchToProps = (dispatch) => bindActionCreators({
  onBaseLayerToggle: setMapBaseLayer,
  onLayerToggle: toggleMapOverlay,
  onLayerVisibilityToggle: toggleMapOverlayVisibility,
  onMapPanelHandleToggle: toggleMapPanelHandle,
  onMapPanelToggle: toggleMapPanel
}, dispatch);

class MapPanelContainer extends React.Component {
  componentDidMount() {
    this.context.store.dispatch(getMapBaseLayers());
    this.context.store.dispatch(getMapLayers());
  }

  componentDidUpdate(prevProps) {
    if (this.props.isMapPanelVisible && prevProps.overlays.length < this.props.overlays.length) {
      const scrollEl = document.querySelector('.map-panel .map-legend');
      if (scrollEl) {
        scrollEl.scrollIntoView({ behavior: 'smooth' });
      }
    }
  }

  render() {
    return (
      <section
        aria-label={this.props.isMapPanelVisible ? 'Kaartlagen legenda, Kaartlagen verbergen' :
          'Kaartlagen legenda, Kaartlagen tonen'}
        aria-expanded={this.props.isMapPanelVisible}
        className={`
          map-panel
          map-panel--${this.props.isMapPanelVisible ? 'expanded' : 'collapsed'}
          map-panel--has${this.props.activeMapLayers.length > 0 ? '' : '-no'}-active-layers
          map-panel--has${this.props.isEachOverlayInvisible ? '-not' : ''}-just-invisible-layers
        `}
      >
        <div className="map-panel__heading">
          <button
            className="map-panel__toggle"
            onClick={this.props.onMapPanelToggle}
            title={this.props.isMapPanelVisible ? 'Kaartlagen verbergen' : 'Kaartlagen tonen'}
          >
            <MapLayersIcon className="map-panel__heading-icon" />
            <h2 className="map-panel__heading-title" aria-hidden="true">Kaartlagen</h2>
            <CollapseIcon className="map-panel__toggle-icon map-panel__toggle-icon--expanded" />
            <ExpandIcon className="map-panel__toggle-icon map-panel__toggle-icon--collapsed" />
          </button>
        </div>
        <div className="scroll-wrapper">
          {this.props.activeMapLayers.length > 0 && (
            <MapLegend
              activeMapLayers={this.props.activeMapLayers}
              onLayerToggle={this.props.onLayerToggle}
              onLayerVisibilityToggle={this.props.onLayerVisibilityToggle}
              overlays={this.props.overlays}
              user={this.props.user}
              zoomLevel={this.props.zoomLevel}
            />
          )}
          <MapPanelHandle
            isMapPanelHandleVisible={this.props.isMapPanelHandleVisible}
            onMapPanelHandleToggle={this.props.onMapPanelHandleToggle}
          >
            <MapType
              activeBaseLayer={this.props.activeBaseLayer}
              baseLayers={this.props.mapBaseLayers}
              onBaseLayerToggle={this.props.onBaseLayerToggle}
            />
            <MapLayers
              activeMapLayers={this.props.activeMapLayers}
              layers={this.props.mapLayers}
              onLayerToggle={this.props.onLayerToggle}
            />
          </MapPanelHandle>
        </div>
      </section>
    );
  }
}

MapPanelContainer.contextTypes = {
  store: PropTypes.object.isRequired
};

MapPanelContainer.defaultProps = {
  activeMapLayers: [],
  atlas: {},
  isMapPanelVisible: false,
  map: {},
  mapBaseLayers: {},
  mapLayers: [],
  user: {},
  zoomLevel: 0
};

MapPanelContainer.propTypes = {
  activeBaseLayer: PropTypes.string.isRequired,
  activeMapLayers: PropTypes.array, // eslint-disable-line
  atlas: PropTypes.object, // eslint-disable-line
  isEachOverlayInvisible: PropTypes.bool.isRequired,
  isMapPanelHandleVisible: PropTypes.bool.isRequired,
  isMapPanelVisible: PropTypes.bool,
  map: PropTypes.object, // eslint-disable-line
  mapBaseLayers: PropTypes.object, // eslint-disable-line
  mapLayers: PropTypes.array, // eslint-disable-line
  onBaseLayerToggle: PropTypes.func, // eslint-disable-line
  onLayerToggle: PropTypes.func, // eslint-disable-line
  onLayerVisibilityToggle: PropTypes.func, // eslint-disable-line
  onMapPanelHandleToggle: PropTypes.func, // eslint-disable-line
  onMapPanelToggle: PropTypes.func.isRequired,
  overlays: PropTypes.array, // eslint-disable-line
  user: PropTypes.object, // eslint-disable-line
  zoomLevel: PropTypes.number
};

export default connect(mapStateToProps, mapDispatchToProps)(MapPanelContainer);
