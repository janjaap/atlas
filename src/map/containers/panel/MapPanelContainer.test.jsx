import React from 'react';
import configureMockStore from 'redux-mock-store';
import { shallow } from 'enzyme';

import MapPanelContainer from './MapPanelContainer';
import MapLayers from '../../components/layers/MapLayers';
import MapLegend from '../../components/legend/MapLegend';
import MapType from '../../components/type/MapType';

describe('MapPanelContainer', () => {
  let store;
  let wrapper;

  beforeEach(() => {
    store = configureMockStore()({
      map: {
        baseLayer: '',
        overlays: []
      },
      mapLayers: [],
      overlays: [{}],
      ui: { isMapPanelHandleVisible: true }
    });
    wrapper = shallow(<MapPanelContainer />, { context: { store } }).dive();
  });

  it('should render MapType and MapLayers', () => {
    expect(wrapper.find(MapType).length).toBe(1);
    expect(wrapper.find(MapLayers).length).toBe(1);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render MapLegend if store contains active map layers', () => {
    expect(wrapper.find(MapLegend).length).toBe(0);
    wrapper.setProps({ activeMapLayers: [{}] });
    expect(wrapper.find(MapLegend).length).toBe(1);
    expect(wrapper).toMatchSnapshot();
  });
});
