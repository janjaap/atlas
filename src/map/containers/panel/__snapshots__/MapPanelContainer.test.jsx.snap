// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapPanelContainer should render MapLegend if store contains active map layers 1`] = `
<section
  aria-expanded={false}
  aria-label="Kaartlagen legenda, Kaartlagen tonen"
  className="
          map-panel
          map-panel--collapsed
          map-panel--has-active-layers
          map-panel--has-not-just-invisible-layers
        "
>
  <div
    className="map-panel__heading"
  >
    <button
      className="map-panel__toggle"
      onClick={[Function]}
      title="Kaartlagen tonen"
    >
      <Component
        className="map-panel__heading-icon"
      />
      <h2
        aria-hidden="true"
        className="map-panel__heading-title"
      >
        Kaartlagen
      </h2>
      <Component
        className="map-panel__toggle-icon map-panel__toggle-icon--expanded"
      />
      <Component
        className="map-panel__toggle-icon map-panel__toggle-icon--collapsed"
      />
    </button>
  </div>
  <div
    className="scroll-wrapper"
  >
    <MapLegend
      activeMapLayers={
        Array [
          Object {},
        ]
      }
      onLayerToggle={[Function]}
      onLayerVisibilityToggle={[Function]}
      overlays={Array []}
      user={Object {}}
      zoomLevel={0}
    />
    <MapPanelHandle
      isMapPanelHandleVisible={true}
      onMapPanelHandleToggle={[Function]}
    >
      <MapType
        activeBaseLayer=""
        baseLayers={Object {}}
        onBaseLayerToggle={[Function]}
      />
      <MapLayers
        activeMapLayers={
          Array [
            Object {},
          ]
        }
        layers={Array []}
        onLayerToggle={[Function]}
      />
    </MapPanelHandle>
  </div>
</section>
`;

exports[`MapPanelContainer should render MapType and MapLayers 1`] = `
<section
  aria-expanded={false}
  aria-label="Kaartlagen legenda, Kaartlagen tonen"
  className="
          map-panel
          map-panel--collapsed
          map-panel--has-no-active-layers
          map-panel--has-not-just-invisible-layers
        "
>
  <div
    className="map-panel__heading"
  >
    <button
      className="map-panel__toggle"
      onClick={[Function]}
      title="Kaartlagen tonen"
    >
      <Component
        className="map-panel__heading-icon"
      />
      <h2
        aria-hidden="true"
        className="map-panel__heading-title"
      >
        Kaartlagen
      </h2>
      <Component
        className="map-panel__toggle-icon map-panel__toggle-icon--expanded"
      />
      <Component
        className="map-panel__toggle-icon map-panel__toggle-icon--collapsed"
      />
    </button>
  </div>
  <div
    className="scroll-wrapper"
  >
    <MapPanelHandle
      isMapPanelHandleVisible={true}
      onMapPanelHandleToggle={[Function]}
    >
      <MapType
        activeBaseLayer=""
        baseLayers={Object {}}
        onBaseLayerToggle={[Function]}
      />
      <MapLayers
        activeMapLayers={Array []}
        layers={Array []}
        onLayerToggle={[Function]}
      />
    </MapPanelHandle>
  </div>
</section>
`;
